core/lib/zksync_core/src/state_keeper/keeper.rs

```diff
@@ -241,6 +225,57 @@ impl ZkSyncStateKeeper {
         Err(Error::Canceled)
     }
 
+    /// This function is meant to be called only once during the state-keeper initialization.
+    /// It will check if we should load a protocol upgrade or a `setChainId` transaction,
+    /// perform some checks and return it.
+    pub(super) async fn load_protocol_upgrade_tx(
+        &mut self,
+        pending_miniblocks: &[MiniblockExecutionData],
+        protocol_version: ProtocolVersionId,
+        l1_batch_number: L1BatchNumber,
+    ) -> Result<Option<ProtocolUpgradeTx>, Error> {
+        // After the Shared Bridge is integrated,
+        // there has to be a setChainId upgrade transaction after the chain genesis.
+        // It has to be the first transaction of the first batch.
+        // The setChainId upgrade does not bump the protocol version, but attaches an upgrade
+        // transaction to the genesis protocol version version.
+        let first_batch_in_shared_bridge =
+            l1_batch_number == L1BatchNumber(1) && !protocol_version.is_pre_shared_bridge();
+        let previous_batch_protocol_version =
+            self.io.load_previous_batch_version_id().await.unwrap();
+
+        let version_changed = protocol_version != previous_batch_protocol_version;
+        let protocol_upgrade_tx = self.io.load_upgrade_tx(protocol_version).await;
+
+        let protocol_upgrade_tx = if pending_miniblocks.is_empty()
+            && (version_changed || first_batch_in_shared_bridge)
+        {
+            tracing::info!("We have upgrade tx to be executed in empty miniblock");
+            protocol_upgrade_tx
+        } else if !pending_miniblocks.is_empty()
+            && (version_changed || first_batch_in_shared_bridge)
+        {
+            // Sanity check: if `txs_to_reexecute` is not empty and upgrade tx is present for this block
+            // then it must be the first one in `txs_to_reexecute`.
+            if protocol_upgrade_tx.is_some() {
+                let first_tx_to_reexecute = &pending_miniblocks[0].txs[0];
+                assert_eq!(
+                    first_tx_to_reexecute.tx_format(),
+                    TransactionType::ProtocolUpgradeTransaction,
+                    "Expected an upgrade transaction to be the first one in pending_miniblocks, but found {:?}",
+                    first_tx_to_reexecute.hash()
+                );
+            }
+            tracing::info!("We have no upgrade tx to execute");
+            None
+        } else {
+            tracing::info!("We are not changing protocol version");
```

Comment:
nit:

```suggestion
            tracing::info!("There is no protocol version change");
```

Also, can you please add more comments to every if clause? what's the difference between no changing protocol version and no ugprade tx to execute?